corehq/apps/userreports/reports/builder/forms.py

```diff
@@ -0,0 +1,569 @@
+                cp: {
+                    'type': 'case_property',
+                    'id': cp,
+                    'column_id': escape_id(cp),
+                    'text': cp,
+                    'source': cp
+                } for cp in self.case_properties
+            }
+
+        if self.source_type == 'form':
+            ret = {}
+            questions = self.source_xform.get_questions([])
+            ret.update({
+                q['value']: {
+                    "type": "question",
+                    "id": q['value'],
+                    "column_id": escape_id(q['value']),
+                    'text': q['label'],
+                    "source": q,
+                } for q in questions
+            })
+            ret.update({
+                p[0]: {
+                    "type": "meta",
+                    "id": p[0],
+                    "column_id": escape_id(p[0]),
+                    'text': p[0],
+                    "source": p,
+                } for p in FORM_METADATA_PROPERTIES
+            })
+            return ret
+
+
+class CreateNewReportForm(forms.Form):
+    """
+    A form for the first page of the report builder.
+    Allows user to specify report type, application, and source.
+    """
+    report_type = forms.ChoiceField(
+        choices=[
+            ('bar_chart', _("Bar Chart")),
+            ('pie_chart', _("Pie Chart")),
+            ('table', _("Table")),
+        ],
+    )
+
+    application = forms.ChoiceField()
+    source_type = forms.ChoiceField(choices=[
+        ("case", _("Case")),
+        ("form", _("Form")),
+    ])
+    report_source = forms.ChoiceField()
+
+    def __init__(self, domain, *args, **kwargs):
+        super(CreateNewReportForm, self).__init__(*args, **kwargs)
+
+        self.domain = domain
+        apps = get_apps_in_domain(domain, full=True, include_remote=False)
+        self.fields['application'].choices = [
+            (app._id, app.name) for app in apps
+        ]
+        self.fields['report_source'].choices = [
+            (ct, ct) for app in apps for ct in app.get_case_types()
+        ] + [
+            (form_id, form_id) for form_id in [
+                form.get_unique_id() for app in apps for form in app.get_forms()
+            ]
+        ]
+
+        self.helper = FormHelper()
+        self.helper.form_class = "form-horizontal"
+        self.helper.form_id = "report-builder-form"
+        self.helper.layout = crispy.Layout(
+            crispy.Fieldset(
+                _('Create New Report'),
+                'report_type',
+                crispy.Field('application', data_bind='value: application'),
+                crispy.Field('source_type', data_bind='value: sourceType'),
+                crispy.Field('report_source', data_bind='''
+                    options: sourcesMap[application()][sourceType()],
+                    optionsText: function(item){return item.text},
+                    optionsValue: function(item){return item.value}
```

Comment:
Added in 84bee6d. Unfortunately I didn't come up with a great way to encapsulate this functionality.